### PR TITLE
Implement Identity API

### LINKED ISSUE #62: As a client I want to be able to update my profile.
In the api/core there should be the possibility to call a function to perform an API call that updates the bio of the user. The user and signal models will have to be updated/created for this. 

This is necessary for #8.

### PR DESCRIPTION
**Pull request by marickmanrho**
Adds:

- Identity model
- Identity$() function to retrieve identity from server
- changeBio() to update user bio.
- changePassword() to update user password.

Does not add:
- changeEmail() functionality to avoid current vulnerability on server side.

Fixes #62

### CODE REVIEW

**Comment by Drvanon** [comment]
Hahaha, all ways pointing to master. Happens to me all the time as well. Will look at this later.

File: gridt/src/app/core/api.service.ts [hunk 1]
```diff
@@ -225,4 +231,42 @@ export class ApiService {
       catchError( this.handleBadAuth() )
     );
   }
+
+   /**
+    * Observable to obtain identity from server
+    */
+
+  public Identity$(): Observable<Identity> {
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.ts]: Because it is an instance of identity, generally these things are set to lower case.

Comment (Drvanon) [reply gridt/src/app/core/api.service.ts]: Also this is a function that takes no input, and returns one specific value, it might be more sensible to make this just a property.

File: gridt/src/app/core/api.service.ts [hunk 2]
```diff
@@ -225,4 +231,42 @@ export class ApiService {
...
+      flatMap((options) => this.http.get<Identity>(
+        `${this.URL}/identity`,
+        options
+      )),
+      catchError( this.handleBadAuth())
+    )
+  }
+
+  public changeBio$( bio: string ) {
+   	console.debug(`Saving new biography to the server. (at least it should now create a http.put)`);
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.ts]: The indentation is not aligned here.

Comment (Drvanon) [reply gridt/src/app/core/api.service.ts]: Also please remove that thing between the parenthesis, it is not useful information.

File: gridt/src/app/core/api.service.ts [hunk 3]
```diff
@@ -225,4 +231,42 @@ export class ApiService {
...
+
+    return this.auth.readyAuthentication$.pipe(
+     flatMap((options) => this.http.put(
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.ts]: Inconsistent indentation.

File: gridt/src/app/core/api.service.ts [hunk 4]
```diff
@@ -225,4 +231,42 @@ export class ApiService {
       catchError( this.handleBadAuth() )
     );
   }
+
+   /**
+    * Observable to obtain identity from server
+    */
+
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.ts]: Unnecessary white space.

File: gridt/src/app/core/api.service.ts [hunk 5]
```diff
@@ -1,12 +1,18 @@
...
 import { Movement } from "./models/movement.model";
 import { User } from './models/user.model';
-import { ServerMessage } from './models/server-responses.model';
+import { Identity } from './models/identity.model';
+
 import { AuthService } from './auth.service';
 
+export interface ServerMessage {
+  message: string;
+}
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.ts]: Why did you redo this?

File: gridt/src/app/core/api.service.spec.ts [hunk 6]
```diff
@@ -86,6 +88,17 @@ describe("ApiService", () => {
...
+
+    mock_identity = [
+      {
+        id: 1,
+        username: "ALittleOne",
+        bio: "Less then one foot tall.",
+        email: "a_little@one.com",
+        avatar: "arandomstring"
+      }
+    ]
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.spec.ts]: Please end with a semicolon.

Comment (Drvanon) [reply gridt/src/app/core/api.service.spec.ts]: Also it is good practice to declare the interfaces it implements.

File: gridt/src/app/core/api.service.spec.ts [hunk 7]
```diff
@@ -253,6 +266,72 @@ describe("ApiService", () => {
...
+      service.Identity$()
+    ).toBeObservable(cold("#", {}, "Can't authenticate: no credentials"));
   });
+
+  it('should fail to update bio when not logged in', () => {
+    authServiceStub.readyAuthentication$ = cold('#', null , "Can't authenticate: no credentials");
+    service = new ApiService(httpClientStub, authServiceStub);
+    let mock_bio = "My very first bio."
+    expect(service.changeBio$( mock_bio )).toBeTruthy( cold('#', null, "Can't authenticate: no credentials"))
+  })
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.spec.ts]: Don't forget the semicolons.

File: gridt/src/app/core/api.service.spec.ts [hunk 8]
```diff
@@ -253,6 +266,72 @@ describe("ApiService", () => {
...
-    )
+    );
+  });
+
+  it('should fail to read identity from server when not logged in', () => {
+    authServiceStub.readyAuthentication$ = cold('#', null , "Can't authenticate: no credentials");
+    service = new ApiService(httpClientStub, authServiceStub);
+
+    expect(
+      service.Identity$()
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.spec.ts]: Semicolon.

File: gridt/src/app/core/api.service.spec.ts [hunk 9]
```diff
@@ -253,6 +266,72 @@ describe("ApiService", () => {
...
+  })
+
+  it('should be able to update the user bio', () => {
+    httpClientStub.put.and.returnValue(of(
+      { "message": "Succesfully updated bio." }
+    ));
+
+    service = new ApiService(httpClientStub, authServiceStub);
+
+    let mock_bio = "My insanely good new bio."
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.spec.ts]: ;

File: gridt/src/app/core/api.service.spec.ts [hunk 10]
```diff
@@ -253,6 +266,72 @@ describe("ApiService", () => {
...
+      default_headers
+    );
+  })
+
+  it('should be able to update the user bio', () => {
+    httpClientStub.put.and.returnValue(of(
+      { "message": "Succesfully updated bio." }
+    ));
+
+    service = new ApiService(httpClientStub, authServiceStub);
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.spec.ts]: Please use dependency injection, I know this can be a quite the pain, but I think it is not honest to circumvent it like this, because there could be things going wrong in this section of the code (e.g. incorrectly configured modules) that would not be caught this way. You can create a new test suite or (and I think this is more beautiful) reset the TestBed using `TestBed.resetTestingModule(); ` that does not solve everything, but it helps.

File: gridt/src/app/core/api.service.spec.ts [hunk 11]
```diff
@@ -253,6 +266,72 @@ describe("ApiService", () => {
...
+
+    expect(
+      service.Identity$()
+    ).toBeObservable(cold("#", {}, "Can't authenticate: no credentials"));
   });
+
+  it('should fail to update bio when not logged in', () => {
+    authServiceStub.readyAuthentication$ = cold('#', null , "Can't authenticate: no credentials");
+    service = new ApiService(httpClientStub, authServiceStub);
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.spec.ts]: Please see the comment on dependency injection.

**Review by Drvanon — CHANGES_REQUESTED** [review]
I would really appreciate it if you could add some comments to the newly created properties and functions in the style of sphinx. I tried to do this previously but it has enough value to push for this I think.

File: gridt/src/app/core/api.service.spec.ts (hunk 6 above)
Comment (marickmanrho) [reply gridt/src/app/core/api.service.spec.ts]: You mean what I do in line 18?

**Comment by marickmanrho** [comment]
It is a mystery to me why the test is passing now... We have to get this sorted asap.

File: gridt/src/app/core/models/identity.model.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
+export interface Identity {
```
Comment (Drvanon) [inline gridt/src/app/core/models/identity.model.ts:1]: Does it make sense to make this a [partial implementation](https://stackoverflow.com/a/57070274/1051438) of User?

File: gridt/src/app/movements/add-movement/add-movement.page.spec.ts [hunk 13]
```diff
@@ -1,6 +1,6 @@
 import { of } from 'rxjs';
```
Comment (Drvanon) [inline gridt/src/app/movements/add-movement/add-movement.page.spec.ts]: This theoretically does not belong in this request.

File: gridt/src/app/movements/movements-detail/movements-detail.page.spec.ts [hunk 14]
```diff
@@ -1,14 +1,21 @@
 import { CUSTOM_ELEMENTS_SCHEMA } from "@angular/core";
```
Comment (Drvanon) [inline gridt/src/app/movements/movements-detail/movements-detail.page.spec.ts]: This technically is outside the scope of the request.

File: gridt/src/app/core/models/identity.model.ts (hunk 12 above)
Comment (marickmanrho) [reply gridt/src/app/core/models/identity.model.ts:1]: It does make sense. Although, if I want to do that I want to exclude the last_signal part. The Identity should be complete for new users and should not require some first interactions with a movement. Hence, in order to do this the User model should be split also. I don't think that is something we want to change now. Maybe an audit of the models in general can be beneficial to the code quality when the settings page is done.

File: gridt/src/app/core/api.service.spec.ts [hunk 15]
```diff
@@ -1,96 +1,155 @@
...
-
+let mock_identity: Identity[] = [
+  {
+    id: 1,
+    username: "ALittleOne",
+    bio: "Less then one foot tall.",
+    email: "a_little@one.com",
+    avatar: "arandomstring"
+  }
+];
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.spec.ts]: This is a list of size 1. Add more identities or make it a single.

File: gridt/src/app/core/api.service.spec.ts [hunk 16]
```diff
@@ -1,96 +1,155 @@
...
-        interval: "daily"
-      }
-    ];
-    mock_subscriptions = [...mock_movements].filter(m => m.subscribed);
-  });
+    service = TestBed.get(ApiService as Type<ApiService>);
 
-  afterEach(() => {
-    httpMock.verify();
+    mock_subscriptions = [...mock_movements].filter(m => m.subscribed);
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.spec.ts:166]: This is an inititialization for the next round, it is more logical to do this in `beforeEach` and remove this function.

File: gridt/src/app/core/api.service.spec.ts [hunk 17]
```diff
@@ -253,6 +294,37 @@ describe("ApiService", () => {
       `${service.URL}/movements/${movement_id}/leader/1`,
       {},
       default_headers
-    )
+    );
+  });
+
+  it('should be able to retreive identity', () => {
+    httpClientStub.get.and.returnValue(of(mock_identity[0]));
```
Comment (Drvanon) [inline gridt/src/app/core/api.service.spec.ts:315]: Theoretically you could also confirm that the right value is URL is called by using `expect(httpClientStub.get).toBeCalledWith({some_object})`. You could do this for the PUT requests as well, if you think it is worth it.

**Review by Drvanon — COMMENTED** [review]
Good work! Please look at my comments and consider them.

File: gridt/src/app/core/api.service.spec.ts (hunk 16 above)
Comment (marickmanrho) [reply gridt/src/app/core/api.service.spec.ts:166]: But it is in beforeEach right now, and indeed the afterEach() function is removed.

**Comment by marickmanrho** [comment]
Due to reverting the changes that where outside the scope of this project the test is now failing. My changes do not cause any errors.

Edit: See also PR #158 which fixes the tests.

**Review by Drvanon — APPROVED** [review]
You can feel free to merge it. Can't do that from phone.
